src/devices/src/virtio/vsock/device.rs

```diff
@@ -65,6 +69,66 @@ pub struct Vsock<B> {
     pub(crate) device_state: DeviceState,
 }
 
+#[derive(Clone, Versionize)]
+pub struct VsockState {
+    cid: u64,
+    queues: Vec<QueueState>,
+    avail_features: u64,
+    acked_features: u64,
+    interrupt_status: usize,
+    activated: bool,
+}
+
+pub struct VsockConstructorArgs<B> {
+    mem: Option<GuestMemoryMmap>,
+    backend: B,
+}
+
+impl<B: VsockBackend> Persist for Vsock<B> {
+    type State = VsockState;
+    type ConstructorArgs = VsockConstructorArgs<B>;
+    type Error = VsockError;
+
+    fn save(self) -> Self::State {
```

Comment:
This cannot consume `self`, saving state is not necessarily a final operation (the microVM should _be able to_ continue running).

It's also very hard to get exclusive ownership of a device (it's wrapped in `Arc` and shared between multiple components of Firecracker).

Reply:
That's right. It just made things super simple later on because I was able to implement `Into` and `From`. For save we shouldn't consume self.